subprojects/execution/src/main/java/org/gradle/internal/execution/steps/SkipEmptyWorkStep.java

```diff
@@ -22,74 +22,215 @@
+        if (outputFilesAfterPreviousExecution.isEmpty()) {
+            LOGGER.info("Skipping {} as it has no source files and no previous output files.", work.getDisplayName());
+            skipOutcome = ExecutionOutcome.SHORT_CIRCUITED;
+        } else {
+            boolean didWork = cleanPreviousTaskOutputs(outputFilesAfterPreviousExecution);
+            if (didWork) {
+                LOGGER.info("Cleaned previous output of {} as it has no source files.", work.getDisplayName());
+                skipOutcome = ExecutionOutcome.EXECUTED_NON_INCREMENTALLY;
+            } else {
+                skipOutcome = ExecutionOutcome.SHORT_CIRCUITED;
+            }
+        }
 
-                    @Override
-                    public ImmutableList<String> getExecutionReasons() {
-                        return ImmutableList.of();
-                    }
+        work.broadcastRelevantFileSystemInputs(skipOutcome);
 
-                    @Override
-                    public Optional<AfterExecutionState> getAfterExecutionState() {
-                        return Optional.empty();
-                    }
+        return new CachingResult() {
+            @Override
+            public Duration getDuration() {
+                return Duration.ZERO;
+            }
 
+            @Override
+            public Try<ExecutionResult> getExecutionResult() {
+                return Try.successful(new ExecutionResult() {
                     @Override
-                    public Optional<OriginMetadata> getReusedOutputOriginMetadata() {
-                        return Optional.empty();
+                    public ExecutionOutcome getOutcome() {
+                        return skipOutcome;
                     }
 
                     @Override
-                    public Duration getDuration() {
-                        return Duration.ZERO;
+                    public Object getOutput() {
+                        return work.loadRestoredOutput(context.getWorkspace());
                     }
-                };
-            })
-            .orElseGet(() -> delegate.execute(work, context));
+                });
+            }
+
+            @Override
+            public CachingState getCachingState() {
+                return CachingState.NOT_DETERMINED;
+            }
+
+            @Override
+            public ImmutableList<String> getExecutionReasons() {
+                return ImmutableList.of();
+            }
+
+            @Override
+            public Optional<AfterExecutionState> getAfterExecutionState() {
+                return Optional.empty();
+            }
+
+            @Override
+            public Optional<OriginMetadata> getReusedOutputOriginMetadata() {
+                return Optional.empty();
+            }
+        };
+    }
+
+    private CachingResult executeWithNoEmptySources(UnitOfWork work, PreviousExecutionContext context, ImmutableSortedMap<String, CurrentFileCollectionFingerprint> newInputFileProperties) {
```

Comment:
I think it would be more readable if we'd only have a method to create a new context with the new inputs, and used the single `executeWithNoEmptySources()` with it.

Reply:
That is what we had before. Seems like I did not understand your previous comment.

Reply:
Ah, I think you got my meaning right. Never mind, I think both solutions are okay.